src/db.rs

```diff
@@ -73,15 +74,19 @@ impl Storage for DB {
             )
         };
         unsafe {
-            go_result.into_ffi_result(err, default)?;
+            if let Err(err) = go_result.into_ffi_result(err, default) {
+                return (Err(err), gas_info);
+            }
         }
 
         if result_buf.ptr.is_null() {
-            return Ok((None, used_gas));
+            return (Ok(None), gas_info);
         }
         // We initialize `result_buf` with a null pointer. if it is not null,
         // that means it was initialized by the go code, with values generated by `memory::allocate_rust`
-        Ok((unsafe { Some(result_buf.consume()) }, 0))
+        let data = unsafe { result_buf.consume() };
+        // TODO: why is this 0?
+        (Ok(Some(data)), GasInfo::with_externally_used(0))
```

Comment:
This makes me think that `GasInfo` should have an `::empty()` or `::none()` constructor for these cases.

Reply:
I considered this too. But then again, should we really design for cases where a method that is suppored to report gas consumtion returns _for free_? Those cases seem to be missing values rather than free on purpose.

Reply:
hmmmm
This line actually does look like a bug!
gas usage was not returned when the value was returned correctly!

in any case, if there will still be cases where we really do want to represent a free action, then i resubmit my suggestion for an empty constructor.

Reply:
> in any case, if there will still be cases where we really do want to represent a free action, then i resubmit my suggestion for an empty constructor.

I agree, like `GasInfo::free()` to shout DO YOU REALLY WANT TO DO THIS?

Reply:
sold :+1:

Reply:
`GasInfo::free()` is now used in the other place where we tend to want 0,0.